lib/zhuyin.ex

```diff
@@ -0,0 +1,376 @@
+    "ㄎ" => "k",
+    "ㄏ" => "h",
+    "ㄐ" => "j",
+    "ㄑ" => "q",
+    "ㄒ" => "x",
+    "ㄓ" => "zh",
+    "ㄔ" => "ch",
+    "ㄕ" => "sh",
+    "ㄖ" => "r",
+    "ㄗ" => "z",
+    "ㄘ" => "c",
+    "ㄙ" => "s"
+  }
+
+  # In pinyin standalone finals are spelled differently than when they are
+  # combined with an initial
+  @standalone_finals %{
+    "ㄧ" => "yi",
+    "ㄨ" => "wu",
+    "ㄩ" => "yu",
+    "ㄧㄚ" => "ya",
+    "ㄨㄚ" => "wa",
+    "ㄧㄥ" => "ying",
+    "ㄧㄤ" => "yang",
+    "ㄧㄝ" => "ye",
+    "ㄨㄛ" => "wo",
+    "ㄨㄥ" => "weng",
+    "ㄨㄤ" => "wang",
+    "ㄧㄠ" => "yao",
+    "ㄨㄞ" => "wai",
+    "ㄩㄝ" => "yue",
+    "ㄩㄥ" => "yong",
+    "ㄧㄡ" => "you",
+    "ㄨㄟ" => "wei",
+    "ㄧㄢ" => "yan",
+    "ㄨㄢ" => "wan",
+    "ㄩㄢ" => "yuan",
+    "ㄧㄣ" => "yin",
+    "ㄨㄣ" => "wen",
+    "ㄩㄣ" => "yun",
+    # Technically standalone initials. Parsed as standalone finals because it's easier to deal with
+    "ㄓ" => "zhi",
+    "ㄔ" => "chi",
+    "ㄕ" => "shi",
+    "ㄖ" => "ri",
+    "ㄗ" => "zi",
+    "ㄘ" => "ci",
+    "ㄙ" => "si",
+    # Standalone finals that are the same in Pinyin as when combined with an initial
+    "ㄦ" => "er",
+    "ㄢ" => "an"
+  }
+
+  @finals %{
+    "ㄧ" => "i",
+    "ㄨ" => "u",
+    "ㄩ" => "v",
+    "ㄚ" => "a",
+    "ㄛ" => "o",
+    "ㄜ" => "e",
+    "ㄝ" => "e",
+    "ㄞ" => "ai",
+    "ㄟ" => "ei",
+    "ㄠ" => "ao",
+    "ㄡ" => "ou",
+    "ㄢ" => "an",
+    "ㄣ" => "en",
+    "ㄤ" => "ang",
+    "ㄥ" => "eng",
+    "ㄦ" => "er",
+    "ㄧㄚ" => "ia",
+    "ㄨㄚ" => "ua",
+    "ㄧㄥ" => "ing",
+    "ㄧㄤ" => "iang",
+    "ㄧㄝ" => "ie",
+    "ㄨㄛ" => "uo",
+    "ㄨㄥ" => "ong",
+    "ㄨㄤ" => "uang",
+    "ㄧㄠ" => "iao",
+    "ㄨㄞ" => "uai",
+    "ㄩㄝ" => "ve",
+    "ㄩㄥ" => "iong",
+    "ㄧㄡ" => "iu",
+    "ㄨㄟ" => "ui",
+    "ㄧㄢ" => "ian",
+    "ㄨㄢ" => "uan",
+    "ㄩㄢ" => "van",
+    "ㄧㄣ" => "in",
+    "ㄨㄣ" => "un",
+    "ㄩㄣ" => "vn"
+  }
+
+  @doc """
+  Create a pinyin struct from a zhuyin struct or list.
+
+  ## Examples
+
+      iex> Zhuyin.to_pinyin(~z/ㄋㄧˇㄏㄠˇ/)
+      ~p/nǐhǎo/
+
+      iex> Zhuyin.to_pinyin(%Zhuyin{initial: "ㄋ", final: "ㄧ", tone: 3})
+      %Pinyin{initial: "n", final: "i", tone: 3}
+
+  """
+  @spec to_pinyin(t() | zhuyin_list()) :: Pinyin.t() | Pinyin.pinyin_list()
+  # Special case for this final and tone combination
+  def to_pinyin(%Zhuyin{initial: "", final: "ㄦ", tone: 0}) do
+    %Pinyin{initial: "", final: "r", tone: "0"}
+  end
+
+  def to_pinyin(zhuyin = %Zhuyin{}) do
+    pinyin_initial = Map.get(@initials, zhuyin.initial, "")
+
+    pinyin_final =
+      if zhuyin.initial == "" do
+        @standalone_finals[zhuyin.final]
+      else
+        @finals[zhuyin.final]
+      end
+
+    %Pinyin{initial: pinyin_initial, final: pinyin_final, tone: zhuyin.tone}
+  end
+
+  def to_pinyin(list) when is_list(list) do
+    list
+    |> Enum.map(fn
+      z = %Zhuyin{} -> Zhuyin.to_pinyin(z)
+      str when is_binary(str) -> str
+    end)
+  end
+
+  @doc """
+  Create a pinyin struct from a zhuyin struct or list.
+
+  ## Examples
```

Comment:
Please use the representation shown in iex (`#Zhuyin<...>`) here to be as close as possible to what the user sees in the shell.

Reply:
I initially used that but it won't work because it's two syllables and looks like this in iex:

```
iex> Zhuyin.from_pinyin(~p/nǐhǎo/)
[#Zhuyin<ㄋㄧˇ>, #Zhuyin<ㄏㄠˇ>]
```

however, mix test has issues parsing that because `#` looks like a comment:

```
  1) doctest Zhuyin.from_pinyin/1 (1) (ZhuyinTest)
     test/zhuyin_test.exs:5
     Doctest did not compile, got: (TokenMissingError) lib/zhuyin.ex:192:2: missing terminator: ] (for "[" starting at line 192)
          |
      192 | [#Zhuyin<ㄋㄧˇ>, #Zhuyin<ㄏㄠˇ>]
          |  ^
     If you are planning to assert on the result of an iex> expression which contains a value inspected as #Name<...>, please make sure the inspected value is placed at the beginning of the expression; otherwise Elixir will treat it as a comment due to the leading sign #.
     doctest:
       iex> Zhuyin.from_pinyin(~p/nǐhǎo/)
       [#Zhuyin<ㄋㄧˇ>, #Zhuyin<ㄏㄠˇ>]
     stacktrace:
       lib/zhuyin.ex:192: Zhuyin (module)
```

What do you think? Should I reduce the example to a single syllable to be able to use the `#Zhuyin<...>` notation?

Reply:
Okay, expanded it. It's also better because it doesn't involve another parsing and makes the test more pure.